worker

```diff
@@ -11,62 +52,96 @@ BH_UUID=${BH_UUID}
-
-#git clone "$GITEA_REPO" dataset && cd dataset # this is standard but clones a full copy of the repo which might be expensive
-#git clone -b "${BH_COMMIT}" --depth 1 "$GITEA_REPO" dataset && cd dataset # this doesn't work unless BH_COMMIT is a branch name
-# this is more complicated but it allows to us to --depth 1 with branches OR commit IDs:
-(mkdir dataset && cd dataset && git init && git remote add origin "$GITEA_REPO" && git fetch --depth 1 origin "${BH_COMMIT}")
-#trap 'echo -n "cleaning "; pwd; chmod -R +w "$WORKDIR" && rm -rf "$WORKDIR"' EXIT
-(
-cd dataset
+# Check out the repository, redirecting all output to stderr.
+(exec 1>&2
+# We'd like to do 'git clone -b "$BH_COMMIT" --depth 1 "$GITEA_REPO"'
+# but maybe "$BH_COMMIT" is a commit ID instead of a branch name, so
+# we do it this way instead.
+echo "# cloning ${GITEA_REPO}"
+git init
+git remote add origin "$GITEA_REPO"
+git fetch --depth 1 origin "$BH_COMMIT"
 git checkout "$BH_COMMIT"
-set -e
-
-if git ls-remote --exit-code "$GITEA_REPO" "refs/heads/git-annex" >/dev/null; then
-  ANNEXED=1
-else
-  ANNEXED=""
-fi
 
-if [ -n "$ANNEXED" ]; then
-  set -x
+# If this is a git-annex repository, we need to get the contents.
+if git ls-remote --exit-code origin refs/heads/git-annex >/dev/null; then
+  echo '# getting git-annexed files'
   # this reduces copies; always overrides annex.hardlink even if that is set system-wide
   git config annex.thin true
   # make sure we don't corrupt origin accidentally
   git config remote.origin.annex-readonly true
   git config annex.private true # XXX this doesn't do anything until git-annex 10
   git annex init
   git annex dead here # this is like annex.private, but has to be run
-  git annex sync --only-annex --no-content # grab the git-annex branch (since we did a shallow clone above)
-  git annex copy --from origin  # NB: using copy --from origin and not git annex get to ensure we're validating the contents of origin and not any special remotes
-else
-  set -x
+  # grab the git-annex branch (since we did a shallow clone above)
+  git annex sync --only-annex --no-content
+  # NB: using 'copy --from origin' and not 'git annex get; to ensure we're
+  # validating the contents of origin and not any special remotes
+  git annex copy --from origin
 fi
-
-bids-validator --verbose .
 )
-) 2>&1
-R=$?
 
-cat <<"EOF"
+echo 1>&2 '# running bids-validator'
+OUTPUT=$(bids-validator .) && STATUS=$? || STATUS=$?
+WARNINGS=$(bids-validator . --json | jq '.issues.warnings | length') || true
+cat 1>&2 <<EOF
+STATUS=${STATUS}
+WARNINGS=${WARNINGS}
+EOF
+
+echo 1>&2 '# formatting output'
+HTML=$(echo "$OUTPUT" | ansifilter --html --fragment --line-numbers --anchors=self)
+
+# Produce the HTML result page on stdout.
+cat <<EOF
+<!DOCTYPE html>
+<html lang=en>
+<title>bids-validator results for ${BH_USER}/${BH_REPO}@${BH_COMMIT}</title>
+<p>Here are the bids-validator results for ${BH_USER}/${BH_REPO}@${BH_COMMIT}:</p>
+<pre>
+${HTML}
 </pre>
 </html>
 EOF
```

Comment:
TIL that bash ignores null bytes when doing variable expansions. At least it prints out a warning that this is the case! I thought with enough quoting it would be safe (apart from the issue that trailing newlines are discarded when doing `$(...)`), but I was insufficiently paranoid. This is a good lesson to learn, thanks.

But since pipes are safe, let's use them, yes. Maybe wrapped in a function, if we want to be nice about formatting the script.